server.go

```diff
@@ -220,6 +220,11 @@ func (s *Server) Close() error {
 		return ErrServerClosed
 	}
 
+	// Stop goroutines.
+	if s.done != nil {
+		close(s.done)
+	}
+
```

Comment:
You need to make sure you unlock the server if you're going to wait for the other goroutines. Otherwise you can deadlock.